### ISSUE TITLE
v0.12.43-debian-cloudwatch and v1.2-debian-cloudwatch failing to start on k8s 1.9.x

### ISSUE THREAD

**Issue by thereverendtom** [fluent/fluentd-kubernetes-daemonset#168]
I can't get the Daemonset to start on my k8s 1.9.3 cluster (generated by kops 1.9.0). This was working previously, but when I redeployed my cluster today it started failing. Each container generates the following error:

```
2018-07-06 01:54:10 +0000 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '1.6.0'
2018-07-06 01:54:10 +0000 [info]: gem 'fluent-plugin-secure-forward' version '0.4.5'
2018-07-06 01:54:10 +0000 [info]: gem 'fluent-plugin-systemd' version '0.0.11'
2018-07-06 01:54:10 +0000 [info]: gem 'fluentd' version '0.12.43'
2018-07-06 01:54:10 +0000 [info]: adding match pattern="fluent.**" type="null"
2018-07-06 01:54:10 +0000 [info]: adding filter pattern="kubernetes.**" type="kubernetes_metadata"
2018-07-06 01:54:10 +0000 [info]: adding match pattern="**" type="cloudwatch_logs"
2018-07-06 01:54:10 +0000 [error]: unexpected error error="No such file or directory - addr"
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/uuidtools-2.1.5/lib/uuidtools.rb:587:in ``'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/uuidtools-2.1.5/lib/uuidtools.rb:587:in `ifconfig'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/uuidtools-2.1.5/lib/uuidtools.rb:634:in `mac_address'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/uuidtools-2.1.5/lib/uuidtools.rb:242:in `block in timestamp_create'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/uuidtools-2.1.5/lib/uuidtools.rb:232:in `synchronize'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/uuidtools-2.1.5/lib/uuidtools.rb:232:in `timestamp_create'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-mixin-config-placeholders-0.4.0/lib/fluent/mixin/config_placeholders.rb:37:in `uuid_timestamp'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-mixin-config-placeholders-0.4.0/lib/fluent/mixin/config_placeholders.rb:80:in `block (2 levels) in configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-mixin-config-placeholders-0.4.0/lib/fluent/mixin/config_placeholders.rb:41:in `block in replace'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-mixin-config-placeholders-0.4.0/lib/fluent/mixin/config_placeholders.rb:41:in `each'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-mixin-config-placeholders-0.4.0/lib/fluent/mixin/config_placeholders.rb:41:in `reduce'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-mixin-config-placeholders-0.4.0/lib/fluent/mixin/config_placeholders.rb:41:in `replace'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-mixin-config-placeholders-0.4.0/lib/fluent/mixin/config_placeholders.rb:96:in `block in configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-mixin-config-placeholders-0.4.0/lib/fluent/mixin/config_placeholders.rb:105:in `configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-plugin-cloudwatch-logs-0.4.5/lib/fluent/plugin/out_cloudwatch_logs.rb:62:in `configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/agent.rb:133:in `add_match'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/agent.rb:64:in `block in configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/agent.rb:57:in `each'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/agent.rb:57:in `configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/root_agent.rb:86:in `configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/engine.rb:129:in `configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/engine.rb:103:in `run_configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/supervisor.rb:489:in `run_configure'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/supervisor.rb:160:in `block in start'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/supervisor.rb:366:in `main_process'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/supervisor.rb:339:in `block in supervise'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/supervisor.rb:338:in `fork'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/supervisor.rb:338:in `supervise'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/supervisor.rb:156:in `start'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/lib/fluent/command/fluentd.rb:173:in `&lt;top (required)&gt;'
  2018-07-06 01:54:10 +0000 [error]: /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
  2018-07-06 01:54:10 +0000 [error]: /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-0.12.43/bin/fluentd:8:in `&lt;top (required)&gt;'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/bin/fluentd:22:in `load'
  2018-07-06 01:54:10 +0000 [error]: /fluentd/vendor/bundle/ruby/2.3.0/bin/fluentd:22:in `&lt;main&gt;'
2018-07-06 01:54:10 +0000 [info]: process finished code=256
2018-07-06 01:54:10 +0000 [warn]: process died within 1 second. exit.
```
My deployment YAML is as follows:
```
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: fluentd
  namespace: kube-system
rules:
  - apiGroups: ["*"]
    resources:
      - pods
      - namespaces
    verbs:
      - get
      - watch
      - list
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: fluentd
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: fluentd-service-account
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: fluentd
subjects:
- kind: ServiceAccount
  name: fluentd
  namespace: kube-system
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: fluentd-logging
  namespace: kube-system
  labels:
    k8s-app: fluentd-logging
    kubernetes.io/cluster-service: "true"
spec:
  selector:
    matchLabels:
      name: fluentd-logging
  template:
    metadata:
      labels:
        name: fluentd-logging
        kubernetes.io/cluster-service: "true"
    spec:
      serviceAccount: fluentd
      serviceAccountName: fluentd
      tolerations:
      - key: node-role.kubernetes.io/master
        effect: NoSchedule
      containers:
      - name: fluentd
        image: fluent/fluentd-kubernetes-daemonset:v0.12-debian-cloudwatch
        env:
        - name: AWS_REGION
          value: us-west-2
        - name: LOG_GROUP_NAME
          valueFrom:
            configMapKeyRef:
              name: cluster-info
              key: cluster-name
        resources:
          limits:
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 200Mi
        volumeMounts:
        - name: varlog
          mountPath: /var/log
        - name: varlibdockercontainers
          mountPath: /var/lib/docker/containers
          readOnly: true
      terminationGracePeriodSeconds: 30
      volumes:
      - name: varlog
        hostPath:
          path: /var/log
      - name: varlibdockercontainers
        hostPath:
          path: /var/lib/docker/containers
```

**Comment by thereverendtom** [comment]
I have also tried updating my deployment to use version 1.2, but I am now getting a completely different error:

```
2018-07-06 06:02:49 +0000 [warn]: #0 [out_cloudwatch_logs] failed to flush the buffer. retry_time=0 next_retry_seconds=2018-07-06 06:02:50 +0000 chunk="5704e663db0ccedabe73f964d668d28f" error_class=Encoding::UndefinedConversionError error="\"\\xC2\" from ASCII-8BIT to UTF-8"
  2018-07-06 06:02:49 +0000 [warn]: #0 /usr/lib/ruby/2.3.0/json/common.rb:224:in `encode'
  2018-07-06 06:02:49 +0000 [warn]: #0 /usr/lib/ruby/2.3.0/json/common.rb:224:in `generate'
  2018-07-06 06:02:49 +0000 [warn]: #0 /usr/lib/ruby/2.3.0/json/common.rb:224:in `generate'
  2018-07-06 06:02:49 +0000 [warn]: #0 /usr/lib/ruby/2.3.0/json/common.rb:395:in `dump'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/out_cloudwatch_logs.rb:217:in `block (2 levels) in write'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/out_cloudwatch_logs.rb:210:in `each'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/out_cloudwatch_logs.rb:210:in `block in write'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/out_cloudwatch_logs.rb:160:in `each'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/out_cloudwatch_logs.rb:160:in `write'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-1.2.2/lib/fluent/plugin/output.rb:1099:in `try_flush'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-1.2.2/lib/fluent/plugin/output.rb:1378:in `flush_thread_run'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-1.2.2/lib/fluent/plugin/output.rb:440:in `block (2 levels) in start'
  2018-07-06 06:02:49 +0000 [warn]: #0 /fluentd/vendor/bundle/ruby/2.3.0/gems/fluentd-1.2.2/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
```
My deployment YAML is the same as above except I have changed the image to `fluent/fluentd-kubernetes-daemonset:v1.2-debian-cloudwatch`.

**Comment by repeatedly** [comment]
See for `addr`: https://github.com/fluent-plugins-nursery/fluent-plugin-cloudwatch-logs/issues/75

**Comment by thereverendtom** [comment]
I did see that! I built a new image that added net-tools, which fixed that issue. Then I was faced with a new error where the container did not have permission to write the ".pos" files to the container log volume mount. I was able to fix that problem by adding the following environment variable to the container:

```
- name: FLUENT_UID
   value: "0"
```
At that point I started getting the `[out_cloudwatch_logs] failed to flush the buffer` error I described above, so I tried upgrading to v1.2 to see if that fixed the issue. It did not, and know I am stuck trying to sort out the buffer issue.
